### PR TITLE
WordSeg model and tests

### PR DESCRIPTION
**Pull request by texasmichelle**
Adds a word segmentation model with tests.

This is an implementation of the paper ["Learning to Discover, Ground, and Use Words with Segmental Neural Language Models"](https://www.aclweb.org/anthology/P19-1645.pdf) by Kazuya Kawakami, Chris Dyer, and Phil Blunsom. This implementation is not affiliated with DeepMind and has not been verified by the authors.

This implementation authored by: @marcrasi @compnerd @saeta

### CODE REVIEW

File: Models/Text/WordSeg/DataSet.swift [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+public struct DataSet {
```
Comment (BradLarson) [inline Models/Text/WordSeg/DataSet.swift:17]: Do we want to move the dataset out into the main datasets (and give it a name other than DataSet) in a follow-on? Related: do we want to start the dataset tests out in the DatasetsTests grouping or move them later?

Comment (texasmichelle) [reply Models/Text/WordSeg/DataSet.swift:17]: Yes, this needs to be rearranged as part of adding the dataset for a full example. This unblocks @asuhan while I put that together.

File: Models/Text/WordSeg/Model.swift [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+
+/// SNLM
+///
+/// A representation of the Segmental Neural Language Model.
+///
+/// \ref https://www.aclweb.org/anthology/P19-1645.pdf
+public struct SNLM: EuclideanDifferentiable, KeyPathIterable {
+  @noDerivative public var conf: Conf
+
+  // MARK: - Encoder
```
Comment (compnerd) [inline Models/Text/WordSeg/Model.swift:55]: Im not sure that these add much value - is this common in the rest of this repository?

File: Models/Text/WordSeg/Model.swift [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+
+  public var embEnc: Embedding<Float>
```
Comment (compnerd) [inline Models/Text/WordSeg/Model.swift:57]: Why not `embeddingEncoder`, `lstmEncoder`?

File: Models/Text/WordSeg/Model.swift [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+
+  public var mlpInterpolation: MLP
+
+  // MARK: - Lexical memory
+
+  public var mlpMemory: MLP
+
+  // MARK: - Character-level decoder
+
+  public var embDec: Embedding<Float>
```
Comment (compnerd) [inline Models/Text/WordSeg/Model.swift:70]: Similarly, why not `embeddingDecoder`, `lstmDecoder`, `denseDecoder`?

File: Models/Text/WordSeg/Model.swift [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+  }
+
+  // MARK: - buildLattice
+
+  //def get_logp_lex(self, logp_lex, candidate):
+  //  if candidate in self.str_vocab:
+  //    candidate_idx = self.str_vocab[candidate]
+  //    return logp_lex[candidate_idx]
+  //  else:
+  //    return torch.log(torch_util.var_from_scaler(0.0, "FloatTensor", self.gpu))
```
Comment (compnerd) [inline Models/Text/WordSeg/Model.swift:192]: Please drop this.  I don't think that we want to have this be in the history.

File: Models/Text/WordSeg/Model.swift [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+
+  func get_logp_lex(_ logp_lex: [Float], _ candidate: CharacterSequence) -> Float {
```
Comment (compnerd) [inline Models/Text/WordSeg/Model.swift:194]: This needs to be renamed for Swift.

File: Models/Text/WordSeg/Model.swift [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+      }
+
+      //# Calculate probabilities
+      //current_state = states[pos]
+      //logg = logg_batch[pos]
+      //logp_lex = logp_lex_batch[pos]
+      //logp_chr = self.decode(candidates, current_state)
+      //# Update semiring score
+      //if pos != 0:
+      //  lattice[pos]["semiring_score"] = semiring.add(lattice[pos]["edges"],
```
Comment (compnerd) [inline Models/Text/WordSeg/Model.swift:232]: Please drop this.

File: Models/Text/WordSeg/Model.swift [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+      //for i, candidate in enumerate(candidates):
+      //  next_pos = pos + len(candidate)
+      //  logp_lex_i = self.get_logp_lex(logp_lex, candidate)
+      //  logp_chr_i = logp_chr[i]
+      //  logp_i = torch_util.logsumexp([logg[0] + logp_lex_i,
+      //                                 logg[1] + logp_chr_i])
+      //  # Create an edge.
+      //  edge = make_edge(pos, next_pos, candidate, logp_i,
+      //                   lattice[pos]["semiring_score"], self.order, self.gpu)
+      //  lattice[next_pos]["edges"].append(edge)
```
Comment (compnerd) [inline Models/Text/WordSeg/Model.swift:258]: Please drop this.

File: Models/Text/WordSeg/SE-0259.swift [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+  ///
+  /// - Parameters:
+  ///   - other: the value to compare with `self`
+  ///   - tolerance: the relative tolerance to use for the comparison.
+  ///     Should be in the range [.ulpOfOne, 1).
+  ///
+  /// - Returns: `true` if `self` is almost equal to `other`; otherwise
+  ///   `false`.
+  @inlinable
+  public func isAlmostEqual(
```
Comment (compnerd) [inline Models/Text/WordSeg/SE-0259.swift:68]: `s/public/internal/` ... just to ensure that we don't accidentally end up with conflicting definitions

File: Models/Text/WordSeg/Vocabularies.swift [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+import TensorFlow
+import ModelSupport
+
+/// Alphabet maps from characters in a string to Int32 representations.
+///
+/// Note: we map from String in order to support multi-character metadata sequences such as </s>.
+///
+/// In Python implementations, this is sometimes called the character vocabulary.
+public struct Alphabet {
```
Comment (compnerd) [inline Models/Text/WordSeg/Vocabularies.swift:23]: Do we need to worry about `Alphabet` being confusing for multiple text models?

File: Tests/TextTests/WordSegmentationTests/ProbeLayers.swift [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+// limitations under the License.
+
+import XCTest
+
+import TensorFlow
+import TextModels
+
+extension SNLM {
+  /// Sets the model parameters to the given parameters exported from the pytorch model.
+  mutating func setTorchParameters(_ p: TorchSNLMParameters) {
```
Comment (compnerd) [inline Tests/TextTests/WordSegmentationTests/ProbeLayers.swift:22]: Can we rename this please?

File: Models/Text/WordSeg/SE-0259.swift (hunk 9 above)
Comment (texasmichelle) [reply Models/Text/WordSeg/SE-0259.swift:68]: This is a bit tricky, given the `@inlinable`.

File: Models/Text/WordSeg/Model.swift [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+// "Learning to Discover, Ground, and Use Words with Segmental Neural Language
+// Models"
+// Kazuya Kawakami, Chris Dyer, Phil Blunsom
+// https://www.aclweb.org/anthology/P19-1645.pdf
+// This implementation is not affiliated with DeepMind and has not been
+// verified by the authors.
+
+import TensorFlow
+
+public struct Conf {
```
[the remark this answers is not in the document]
Comment (texasmichelle) [reply Models/Text/WordSeg/Model.swift]: Done

File: Tests/TextTests/WordSegmentationTests/ProbeLayers.swift (hunk 11 above)
Comment (texasmichelle) [reply Tests/TextTests/WordSegmentationTests/ProbeLayers.swift]: Done

File: Models/Text/WordSeg/Model.swift (hunk 8 above)
Comment (texasmichelle) [reply Models/Text/WordSeg/Model.swift:274]: Done

File: Models/Text/WordSeg/Model.swift (hunk 7 above)
Comment (texasmichelle) [reply Models/Text/WordSeg/Model.swift:248]: Done

File: Models/Text/WordSeg/DataSet.swift (hunk 1 above)
Comment (texasmichelle) [reply Models/Text/WordSeg/DataSet.swift]: Marked as internal

**Comment by texasmichelle** [comment]
OK, most of @compnerd's comments have been addressed and I created #499 for scope cleanup.

File: Models/Text/WordSeg/Lattice.swift [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+#elseif os(Windows)
+  import ucrt
+#else
+  import Glibc
+#endif
+
+/// Lattice
+///
+/// Represents the lattice used by the WordSeg algorithm.
+public struct Lattice: Differentiable {
```
[the remark this answers is not in the document]
Comment (texasmichelle) [reply Models/Text/WordSeg/Lattice.swift:28]: Changing this scope looks to be a bit more involved. Punting to #499

File: Models/Text/WordSeg/DataSet.swift (hunk 1 above)
Comment (texasmichelle) [reply Models/Text/WordSeg/DataSet.swift]: #500 for adding a dataset and cleaning this up.

**Pull request merged**